sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineResponse, PipelineRequest, SendRequest } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { delay } from "../util/helpers";
+import { createClientLogger } from "@azure/logger";
+import { tryCreateSpan, tryTraceError, tryTraceResponse } from "./tracingPolicy";
+import { Span } from "@azure/core-tracing";
+import { getUserAgentValue } from "../util/userAgent";
+import { RetryStrategy, RetryStrategyState } from "../retryStrategies/retryStrategy";
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(...strategies: RetryStrategy[]): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
```

Comment:
My original hunch was that policies might need to be reordered so that tracingPolicy comes after retry policy - I have honestly not looked too closely at the pipeline policies (haven't had to) but I was wondering if other folks have any intuition here:

do you think that by moving `tracingPolicy` around in the pipeline ordering we can avoid having tracing code in the retry policy?

Reply:
@maorleger great point! I’ll try it out and update the PR if that makes sense

Reply:
If I do as follows in this test:

```ts
      const testRetryPolicy = retryPolicy({
        name: "testRetryStrategy",
        meetsConditions({ responseError }) {
          return Boolean(responseError && responseError!.code === "ENOENT");
        },
        updateRetryState(state: RetryStrategyState): RetryStrategyState {
          state.retryAfterInMs = 100;
          return state;
        }
      });

      const pipeline = createEmptyPipeline();
      pipeline.addPolicy(testRetryPolicy, { phase: "Retry" });
      pipeline.addPolicy(tracingPolicy());

      const replies = [testError, successResponse];
      const testHttpClient: HttpClient = {
        async sendRequest() {
          const response = replies.shift()!;
          if ((response as RestError).name === "RestError") {
            throw response;
          }
          return response as PipelineResponse;
        }
      };
      const result = await pipeline.sendRequest(testHttpClient, request);
```

I only get one started span when I check `mockTracer.getStartedSpans()`. Would that align with what you expect? I wanted to see more spans based on each retry.

Your feedback is appreciated!

Reply:
I see - I would expect N spans to be created for N retries (1 per retry / http call). I am just guessing but likely you are only seeing one span because the policies are still being reordered since retryPolicy is in the `retry` phase - so it's likely still coming after `tracingPolicy`.

Maybe @xirzec has an idea for an ordering that makes sense here? Original Q: do you think that by moving tracingPolicy around in the pipeline ordering we can avoid having tracing code in the retry policy?

Reply:
ohh adding it to the retry phase seems to do the work! I’ll use that.

Reply:
here it is: https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/e6386bc4f0b31f22de2dbbcdee07545334058e0e how does it look?

Reply:
I think it's so much nicer this way if it works and doesn't have any other impact on non-retryable scenarios but would definitely want one of the pipeline experts to chime in. Thank you for making the change!

Reply:
@maorleger thank you for your help! ❤️